compiler/ballerina-lang/src/main/java/io/ballerina/projects/CompilationOptions.java

```diff
@@ -62,10 +64,21 @@ public class CompilationOptions {
         this.exportOpenAPI = exportOpenAPI;
         this.exportComponentModel = exportComponentModel;
         this.enableCache = enableCache;
+        if (CompilationOptions.optimizeCodegen == null) {
+            CompilationOptions.optimizeCodegen = optimizeCodegen;
+        }
+        if (CompilationOptions.optimizeReport == null) {
+            CompilationOptions.optimizeReport = optimizeReport;
+        }
```

Comment:
If possible let's move this logic to the factory method and have it pass the correct value in via the constructor, which we can store in the instance as a immutable field. This means we need to figure out a way to get rid of `ressetStaticCompilationOptions`, which I feel not a good idea anyway because it allows people to mutate the state of every CompilationOption from any arbitery point.

 Also since we are modifying this anyway let's see if we can make the consturctor private as well. It is not a good idea to have both a accessible constructor and factory method

Reply:
After an offline discussion with @heshanpadmasiri, we decided to move the static variables to the builder class. And for resetting the static variables after each related unit test, we decided to use reflection. Changes can be found in ca52755bff1c0ddaeb396ed614106173315fdaa7.